provider/adapters/src/fork/by_error.rs

```diff
@@ -15,12 +13,15 @@ use icu_provider::prelude::*;
 /// This is an abstract forking provider that must be provided with a type implementing the
 /// [`ForkByErrorPredicate`] trait.
 #[derive(Debug, PartialEq, Eq)]
-pub struct ForkByErrorProvider<P0, P1, F>(pub P0, pub P1, PhantomData<F>);
+pub struct ForkByErrorProvider<P0, P1, F>(pub P0, pub P1, F);
 
 impl<P0, P1, F> ForkByErrorProvider<P0, P1, F> {
     /// Create a new provider that forks between the two children.
-    pub fn new(p0: P0, p1: P1) -> Self {
-        Self(p0, p1, PhantomData)
+    ///
+    /// The `predicate` argument should be an instance of a struct implementing
+    /// [`ForkByErrorPredicate`].
```

Comment:
So...let's add a type bound?

Reply:
Well, I don't _need_ the type bound in this function. I only like adding the type bound if I'm actually using the bound in this particular function. If someone passes something not implementing the trait to this function, then they will get errors when they try using the `BufferProvider` and other provider impls.

For example, if I pass `()` as the predicate, the error looks like this:

```
error[E0599]: the method `load_buffer` exists for struct `ForkByErrorProvider<..., ()>`, but its trait bounds were not satisfied
...
note: trait bound `(): ForkByErrorPredicate` was not satisfied
  --> provider/adapters/src/fork/by_error.rs:41:8
   |
37 | impl<P0, P1, F> BufferProvider for ForkByErrorProvider<P0, P1, F>
   |                 --------------     ------------------------------
...
41 |     F: ForkByErrorPredicate,
   |        ^^^^^^^^^^^^^^^^^^^^ unsatisfied trait bound introduced here

```

Reply:
Which is a lot more confusing. If you're putting the bound in the documentation, you might as well make the compiler enforce it

Reply:
@Manishearth can you break the tie? :)

Reply:
In theory, there may be a specific impl that doesn't actually use the predicate. In such cases, maybe you don't want to provide a predicate. So I still believe that it is not correct to impose the bound on this function.

Reply:
But there isn't. And if you come up with one, removing the bound is non-breaking

Reply:
It seems that your position is that we should add the bound as a way to help developers, almost like better documentation. My position is that bounds should only be added when they are strictly necessary.

@Manishearth?

Reply:
Established Rust style is "don't add unnecessary bounds to methods that don't need them, even if the the type is useless without those bounds"

There are reasons for this: mostly the unnecessary bounds can become infectuous especially when on structs.

For example, [`HashMap::new()`](https://doc.rust-lang.org/stable/std/collections/struct.HashMap.html#method.new) has no bounds  even though a hashmap without `K: Hash` is completely useless.

Reply:
Also, this function is named `new_with_predicate`, which strongly suggests to the reader that the argument should indeed be a predicate function, with the signature of the predicate function listed in the docs for convenience.

Reply:
Fine